src/libstd/io/buffered.rs

```diff
@@ -818,6 +899,279 @@ impl<W> fmt::Display for IntoInnerError<W> {
+/// enables Stdout to be alternately in line-buffered or block-buffered mode.
+#[derive(Debug)]
+pub(super) struct LineWriterShim<'a, W: Write> {
+    buffer: &'a mut BufWriter<W>,
+}
+
+impl<'a, W: Write> LineWriterShim<'a, W> {
+    pub fn new(buffer: &'a mut BufWriter<W>) -> Self {
+        Self { buffer }
+    }
+
+    /// Get a reference to the inner writer (that is, the writer wrapped by
+    /// the BufWriter)
+    fn inner(&self) -> &W {
+        self.buffer.get_ref()
+    }
+
+    /// Get a mutable reference to the inner writer (that is, the writer
+    /// wrapped by the BufWriter). Be careful with this writer, as writes to
+    /// it will bypass the buffer.
+    fn inner_mut(&mut self) -> &mut W {
+        self.buffer.get_mut()
+    }
+
+    /// Get the content currently buffered in self.buffer
+    fn buffered(&self) -> &[u8] {
+        self.buffer.buffer()
+    }
+}
+
+impl<'a, W: Write> Write for LineWriterShim<'a, W> {
+    /// Write some data into this BufReader with line buffering. This means
+    /// that, if any newlines are present in the data, the data up to the last
+    /// newline is sent directly to the underlying writer, and data after it
+    /// is buffered. Returns the number of bytes written.
+    ///
+    /// This function operates on a "best effort basis"; in keeping with the
+    /// convention of `Write::write`, it makes at most one attempt to write
+    /// new data to the underlying writer. If that write only reports a partial
+    /// success, the remaining data will be buffered.
+    ///
+    /// Because this function attempts to send completed lines to the underlying
+    /// writer, it will also flush the existing buffer if it contains any
+    /// newlines, even if the incoming data does not contain any newlines.
+    fn write(&mut self, buf: &[u8]) -> io::Result<usize> {
+        let newline_idx = match memchr::memrchr(b'\n', buf) {
+            // If there are no new newlines (that is, if this write is less than
+            // one line), just do a regular buffered write
+            None => {
+                // Check for prior partial line writes that need to be retried.
+                // Only retry if the buffer contains a completed line, to
+                // avoid flushing partial lines.
+                if let Some(b'\n') = self.buffered().last().copied() {
+                    self.buffer.flush_buf()?;
+                }
+                return self.buffer.write(buf);
+            }
+            // Otherwise, arrange for the lines to be written directly to the
+            // inner writer.
+            Some(newline_idx) => newline_idx + 1,
+        };
+
+        // Flush existing content to prepare for our write
+        self.buffer.flush_buf()?;
+
+        // This is what we're going to try to write directly to the inner
+        // writer. The rest will be buffered, if nothing goes wrong.
+        let lines = &buf[..newline_idx];
```

Comment:
I believe we can just write the entire `buf` directly to the inner writer. AFAIK `LineWriter` doesn't make any guarantees that partial lines are actually buffered, it only guarantees that newlines will be written out immediately instead of being buffered.

Reply:
When there are multiple concurrent writers to a pipe it is still useful to buffer up whole lines to avoid interleaving text. This is a common problem with multithreaded programs printing to stdout.

Reply:
I strongly considered that. However, in my opinion, taking steps to only write whole lines is the least surprising behavior; my expectation as a client of this API would be that (outside of very-long lines) it only flushes whole lines, regardless of the shape of the input.